modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestPropagationSecurityContextTest.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.ignite.internal.processors.security.impl.TestSecurityData;
+import org.apache.ignite.internal.processors.security.impl.TestSecurityPluginProvider;
+import org.apache.ignite.internal.processors.security.impl.TestSecurityProcessor;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.internal.visor.query.VisorQueryTask;
+import org.apache.ignite.internal.visor.query.VisorQueryTaskArg;
+import org.apache.ignite.plugin.security.SecurityPermissionSetBuilder;
+import org.junit.Test;
+
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_BASELINE_AUTO_ADJUST_ENABLED;
+import static org.apache.ignite.configuration.WALMode.NONE;
+import static org.apache.ignite.plugin.security.SecurityPermission.TASK_EXECUTE;
+import static org.apache.ignite.plugin.security.SecurityPermissionSetBuilder.ALLOW_ALL;
+
+/**
+ * Test REST with enabled authentication.
+ */
+public class JettyRestPropagationSecurityContextTest extends JettyRestProcessorCommonSelfTest {
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        System.setProperty(IGNITE_BASELINE_AUTO_ADJUST_ENABLED, "false");
+
+        U.resolveWorkDirectory(U.defaultWorkDirectory(), "db", true);
+
+        super.beforeTestsStarted();
+
+        // We need to activate cluster.
+        grid(0).cluster().active(true);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTestsStopped() throws Exception {
+        super.afterTestsStopped();
+
+        System.clearProperty(IGNITE_BASELINE_AUTO_ADJUST_ENABLED);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected String signature() {
+        return null;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        DataStorageConfiguration dsCfg = new DataStorageConfiguration()
+            .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                .setMaxSize(100 * 1024 * 1024)
+                .setPersistenceEnabled(true))
+            .setWalMode(NONE);
+
+        cfg.setDataStorageConfiguration(dsCfg)
+           .setAuthenticationEnabled(true)
+            .setPluginProviders(new TestSecurityPluginProvider("server", "server", ALLOW_ALL, false,
+                new TestSecurityData("client", "client", SecurityPermissionSetBuilder.create()
+                    .defaultAllowAll(false)
+                    .appendTaskPermissions("org.apache.ignite.internal.visor.query.VisorQueryTask", TASK_EXECUTE)
+                    .appendTaskPermissions("org.apache.ignite.internal.visor.compute.VisorGatewayTask", TASK_EXECUTE)
+                    .build()
+                )));
+
+        return cfg;
+    }
+
+    /**
+     *
+     */
+    @Test
+    public void restClientShouldPropagateSecurityContext() throws Exception {
```

Comment:
I think the best way to test visor tasks is to have two clients: the first client does have permission, and the second doesn't.
If you run a task on behalf of the first client, you get cache data.
If you run a task as the second client, you should get an exception.

Reply:
Yes, but it's a big integration test which requires additional time for creating. Here committed the reproducer. It checks that a context propagation works fine. We have tests that check permissions. If use transitive relation rule, it enough.